zero/src/rpc/jumpdest.rs

```diff
@@ -0,0 +1,512 @@
+            disable_memory: Some(!memory),
+            disable_storage: Some(!storage),
+            ..GethDefaultTracingOptions::default()
+        },
+        tracer: None,
+        ..GethDebugTracingOptions::default()
+    }
+}
+
+/// Predicate that determines whether a `StructLog` that includes memory is
+/// required.
+fn trace_contains_create(structlog: &[StructLog]) -> bool {
+    structlog
+        .iter()
+        .any(|entry| entry.op == "CREATE" || entry.op == "CREATE2")
+}
+
+/// Gets the lightest possible structlog for transcation `tx_hash`.
+pub(crate) async fn get_normalized_structlog<ProviderT, TransportT>(
+    provider: &ProviderT,
+    tx_hash: &B256,
+) -> Result<Option<Vec<StructLog>>, RpcError<TransportErrorKind>>
+where
+    ProviderT: Provider<TransportT>,
+    TransportT: Transport + Clone,
+{
+    let inner = async {
+        // Optimization: It may be a better default to pull the stack immediately.
+        let stackonly_structlog_trace = provider
+            .debug_trace_transaction(*tx_hash, structlog_tracing_options(true, false, false))
+            .await?;
+
+        let stackonly_structlog_opt: Option<Vec<StructLog>> =
+            normalize_structlog(&stackonly_structlog_trace).await;
+
+        let need_memory = stackonly_structlog_opt
+            .as_deref()
+            .is_some_and(trace_contains_create);
+        trace!("Need structlog with memory: {need_memory}");
+
+        if need_memory.not() {
+            return Ok(stackonly_structlog_opt);
+        };
+
+        let memory_structlog_fut = provider.debug_trace_transaction(
+            *tx_hash,
+            structlog_tracing_options(true, need_memory, false),
+        );
+
+        let memory_structlog = normalize_structlog(&memory_structlog_fut.await?).await;
+
+        Ok::<Option<Vec<_>>, RpcError<TransportErrorKind>>(memory_structlog)
+    };
+
+    match timeout(TIMEOUT_LIMIT, inner).await {
+        Err(ellapsed_error) => Err(RpcError::Transport(TransportErrorKind::Custom(Box::new(
+            ellapsed_error,
+        )))),
+        Ok(structlog_res) => Ok(structlog_res?),
+    }
+}
+
+/// Generate at JUMPDEST table by simulating the call stack in EVM,
+/// using a Geth structlog as input.
+pub(crate) fn generate_jumpdest_table(
+    tx: &Transaction,
+    struct_log: &[StructLog],
+    tx_traces: &BTreeMap<Address, TxnTrace>,
+) -> anyhow::Result<(JumpDestTableWitness, CodeDb)> {
+    trace!("Generating JUMPDEST table for tx: {}", tx.hash);
+
+    let mut jumpdest_table = JumpDestTableWitness::default();
+    let mut code_db = CodeDb::default();
+
+    // This map does not contain `initcodes`.
+    let callee_addr_to_code_hash: HashMap<Address, H256> = tx_traces
+        .iter()
+        .map(|(callee_addr, trace)| (callee_addr, &trace.code_usage))
+        .filter(|(_callee_addr, code_usage)| code_usage.is_some())
+        .map(|(callee_addr, code_usage)| {
+            (*callee_addr, code_usage.as_ref().unwrap().get_code_hash())
```

Comment:
Please use `filter_map` rather than checking then unwrapping

Reply:
Good catch. Isn't there usually a clippy hint for this?